vllm/v1/worker/gpu_model_runner.py

```diff
@@ -945,23 +945,33 @@ def execute_model(
 
         # Sample the next token and get logprobs if needed.
         sampling_metadata = self.input_batch.sampling_metadata
-        if not self.use_spec_decode:
-            sampler_output = self.model.sample(
-                logits=logits,
-                sampling_metadata=sampling_metadata,
-            )
-        else:
-            target_probs = self.model.sampler.compute_probs(
-                logits, sampling_metadata)
+
+        if self.use_spec_decode:
             scheduled_request_ids = scheduler_output.num_scheduled_tokens.keys(
             )
             draft_token_ids = [
                 scheduler_output.scheduled_spec_decode_tokens.get(req_id, [])
                 for req_id in scheduled_request_ids
             ]
-            sampler_output = self.rejection_sampler(draft_token_ids,
-                                                    target_probs,
-                                                    sampling_metadata)
+            sample_lens = [len(tokens) + 1 for tokens in draft_token_ids]
+            recover_logits_idx = np.cumsum(sample_lens) - 1
+            target_probs = self.model.sampler.compute_probs(
+                logits, sampling_metadata, sample_lens)
+            bonus_token_ids = self.model.sample(
+                logits=logits[recover_logits_idx, :],
+                sampling_metadata=sampling_metadata,
+            ).sampled_token_ids
+            sampler_output = self.rejection_sampler(
```

Comment:
Does it also support logprobs/prompt logprobs? I think we should reuse some values from the output of `self.model.sample` above.